// @magicblock-labs/ephemeral-validator/ephemeralValidator.js
#!/usr/bin/env node
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = __importDefault(require("fs"));
const child_process_1 = require("child_process");
const path_1 = __importDefault(require("path"));
const os_1 = require("os");
const getVersions_1 = require("./getVersions");
const PACKAGE_VERSION = `ephemeral-validator ${getVersions_1.VERSIONS.EPHEMERAL_VALIDATOR}`;
function getBinaryVersion(location) {
    const result = (0, child_process_1.spawnSync)(location, ["--version"]);
    const error = (result.error && result.error.toString()) ||
        (result.stderr.length > 0 && result.stderr.toString().trim()) ||
        null;
    return [error, result.stdout && result.stdout.toString().trim()];
}
function getExePath() {
    let os = (0, os_1.platform)();
    let extension = "";
    if (["win32", "cygwin"].includes(os)) {
        os = "windows";
        extension = ".exe";
    }
    const binaryName = `@magicblock-labs/ephemeral-validator-${os}-${(0, os_1.arch)()}/bin/ephemeral-validator${extension}`;
    try {
        return require.resolve(binaryName);
    }
    catch (e) {
        throw new Error(`Couldn't find application binary inside node_modules for ${os}-${(0, os_1.arch)()}, expected location: ${binaryName}`);
    }
}
function runWithForwardedExit(child) {
    child.on("exit", (code, signal) => {
        process.on("exit", () => {
            if (signal) {
                process.kill(process.pid, signal);
            }
            else if (code !== null) {
                process.exit(code);
            }
        });
    });
    process.on("SIGINT", () => {
        child.kill("SIGINT");
        child.kill("SIGTERM");
    });
}
function runEphemeralValidator(location) {
    const args = process.argv.slice(2);
    const env = Object.assign(Object.assign({}, process.env), { RUST_LOG: "quiet" });
    const ephemeralValidator = (0, child_process_1.spawn)(location, args, { stdio: "inherit", env });
    runWithForwardedExit(ephemeralValidator);
}
function tryPackageEphemeralValidator() {
    try {
        const path = getExePath();
        runEphemeralValidator(path);
        return true;
    }
    catch (e) {
        console.error("Failed to run bolt from package:", e instanceof Error ? e.message : e);
        return false;
    }
}
function trySystemEphemeralValidator() {
    var _a;
    const absolutePath = (_a = process.env.PATH) === null || _a === void 0 ? void 0 : _a.split(path_1.default.delimiter).filter((dir) => dir !== path_1.default.dirname(process.argv[1])).find((dir) => {
        try {
            fs_1.default.accessSync(`${dir}/ephemeral-validator`, fs_1.default.constants.X_OK);
            return true;
        }
        catch (_a) {
            return false;
        }
    });
    if (!absolutePath) {
        console.error(`Could not find globally installed ephemeral-validator, please install with cargo.`);
        process.exit(1);
    }
    const absoluteBinaryPath = `${absolutePath}/ephemeral-validator`;
    const [error, binaryVersion] = getBinaryVersion(absoluteBinaryPath);
    if (error !== null) {
        console.error(`Failed to get version of global binary: ${error}`);
        return;
    }
    if (binaryVersion !== PACKAGE_VERSION) {
        console.error(`Globally installed ephemeral-validator version is not correct. Expected "${PACKAGE_VERSION}", found "${binaryVersion}".`);
        return;
    }
    runEphemeralValidator(absoluteBinaryPath);
}
// If the first argument is our special command, run the test validator and exit.
tryPackageEphemeralValidator() || trySystemEphemeralValidator();
