// @magicblock-labs/ephemeral-rollups-sdk/lib/instructions/ephemeral-spl-token-program/ephemeralAta.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.decodeEphemeralAta = decodeEphemeralAta;
exports.encodeEphemeralAta = encodeEphemeralAta;
exports.decodeGlobalVault = decodeGlobalVault;
exports.encodeGlobalVault = encodeGlobalVault;
exports.deriveEphemeralAta = deriveEphemeralAta;
exports.deriveVault = deriveVault;
exports.deriveVaultAta = deriveVaultAta;
exports.deriveShuttleEphemeralAta = deriveShuttleEphemeralAta;
exports.deriveShuttleAta = deriveShuttleAta;
exports.deriveShuttleWalletAta = deriveShuttleWalletAta;
exports.initEphemeralAtaIx = initEphemeralAtaIx;
exports.initVaultAtaIx = initVaultAtaIx;
exports.initVaultIx = initVaultIx;
exports.transferToVaultIx = transferToVaultIx;
exports.depositSplTokensIx = depositSplTokensIx;
exports.delegateEphemeralAtaIx = delegateEphemeralAtaIx;
exports.initShuttleEphemeralAtaIx = initShuttleEphemeralAtaIx;
exports.delegateShuttleEphemeralAtaIx = delegateShuttleEphemeralAtaIx;
exports.mergeShuttleIntoAtaIx = mergeShuttleIntoAtaIx;
exports.undelegateAndCloseShuttleEphemeralAtaIx = undelegateAndCloseShuttleEphemeralAtaIx;
exports.withdrawSplIx = withdrawSplIx;
exports.undelegateIx = undelegateIx;
exports.createEataPermissionIx = createEataPermissionIx;
exports.resetEataPermissionIx = resetEataPermissionIx;
exports.delegateEataPermissionIx = delegateEataPermissionIx;
exports.undelegateEataPermissionIx = undelegateEataPermissionIx;
exports.delegateSpl = delegateSpl;
exports.delegateSplIdempotent = delegateSplIdempotent;
exports.delegatePrivateSpl = delegatePrivateSpl;
const web3_js_1 = require("@solana/web3.js");
const constants_js_1 = require("../../constants.js");
const pda_js_1 = require("../../pda.js");
const DEFAULT_PRIVATE_VALIDATOR = new web3_js_1.PublicKey("FnE6VJT5QNZdedZPnCoLsARgBwoE6DeJNjBs2H1gySXA");
const TOKEN_PROGRAM_ID = new web3_js_1.PublicKey("TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA");
const ASSOCIATED_TOKEN_PROGRAM_ID = new web3_js_1.PublicKey("ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL");
function getAssociatedTokenAddressSync(mint, owner, allowOwnerOffCurve = false, programId = TOKEN_PROGRAM_ID, associatedTokenProgramId = ASSOCIATED_TOKEN_PROGRAM_ID) {
    if (!allowOwnerOffCurve && !web3_js_1.PublicKey.isOnCurve(owner)) {
        throw new Error("Owner public key is off-curve");
    }
    const [ata] = web3_js_1.PublicKey.findProgramAddressSync([owner.toBuffer(), programId.toBuffer(), mint.toBuffer()], associatedTokenProgramId);
    return ata;
}
function createAssociatedTokenAccountIdempotentInstruction(payer, associatedToken, owner, mint, programId = TOKEN_PROGRAM_ID, associatedTokenProgramId = ASSOCIATED_TOKEN_PROGRAM_ID) {
    const data = Buffer.from([1]);
    return new web3_js_1.TransactionInstruction({
        programId: associatedTokenProgramId,
        keys: [
            { pubkey: payer, isSigner: true, isWritable: true },
            { pubkey: associatedToken, isSigner: false, isWritable: true },
            { pubkey: owner, isSigner: false, isWritable: false },
            { pubkey: mint, isSigner: false, isWritable: false },
            { pubkey: web3_js_1.SystemProgram.programId, isSigner: false, isWritable: false },
            { pubkey: programId, isSigner: false, isWritable: false },
        ],
        data,
    });
}
function decodeEphemeralAta(info) {
    if (info.data.length < 72) {
        throw new Error("Invalid EphemeralAta account data length");
    }
    const owner = new web3_js_1.PublicKey(info.data.subarray(0, 32));
    const mint = new web3_js_1.PublicKey(info.data.subarray(32, 64));
    const amount = BigInt(info.data.readBigUInt64LE(64));
    return {
        owner,
        mint,
        amount,
    };
}
function encodeEphemeralAta(eata) {
    const buffer = Buffer.alloc(72);
    buffer.set(eata.owner.toBytes(), 0);
    buffer.set(eata.mint.toBytes(), 32);
    buffer.writeBigUInt64LE(eata.amount, 64);
    return buffer;
}
function decodeGlobalVault(info) {
    if (info.data.length < 32) {
        throw new Error("Invalid GlobalVault account data length");
    }
    const mint = new web3_js_1.PublicKey(info.data.subarray(0, 32));
    return { mint };
}
function encodeGlobalVault(vault) {
    const buffer = Buffer.alloc(32);
    buffer.set(vault.mint.toBytes(), 0);
    return buffer;
}
function deriveEphemeralAta(owner, mint) {
    return web3_js_1.PublicKey.findProgramAddressSync([owner.toBuffer(), mint.toBuffer()], constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID);
}
function deriveVault(mint) {
    return web3_js_1.PublicKey.findProgramAddressSync([mint.toBuffer()], constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID);
}
function deriveVaultAta(mint, vault) {
    return getAssociatedTokenAddressSync(mint, vault, true);
}
function deriveShuttleEphemeralAta(owner, mint, shuttleId) {
    if (!Number.isInteger(shuttleId) ||
        shuttleId < 0 ||
        shuttleId > 4294967295) {
        throw new Error("shuttleId must fit in u32");
    }
    const shuttleIdSeed = Buffer.alloc(4);
    shuttleIdSeed.writeUInt32LE(shuttleId, 0);
    return web3_js_1.PublicKey.findProgramAddressSync([owner.toBuffer(), mint.toBuffer(), shuttleIdSeed], constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID);
}
function deriveShuttleAta(shuttleEphemeralAta, mint) {
    return web3_js_1.PublicKey.findProgramAddressSync([shuttleEphemeralAta.toBuffer(), mint.toBuffer()], constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID);
}
function deriveShuttleWalletAta(mint, shuttleEphemeralAta) {
    return getAssociatedTokenAddressSync(mint, shuttleEphemeralAta, true);
}
function initEphemeralAtaIx(ephemeralAta, owner, mint, payer, bump) {
    return new web3_js_1.TransactionInstruction({
        programId: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
        keys: [
            { pubkey: ephemeralAta, isSigner: false, isWritable: true },
            { pubkey: payer, isSigner: true, isWritable: true },
            { pubkey: owner, isSigner: false, isWritable: false },
            { pubkey: mint, isSigner: false, isWritable: false },
            { pubkey: web3_js_1.SystemProgram.programId, isSigner: false, isWritable: false },
        ],
        data: Buffer.from([0, bump]),
    });
}
function initVaultAtaIx(payer, vaultAta, vault, mint) {
    return createAssociatedTokenAccountIdempotentInstruction(payer, vaultAta, vault, mint);
}
function initVaultIx(vault, mint, payer, bump) {
    const vaultAta = deriveVaultAta(mint, vault);
    return new web3_js_1.TransactionInstruction({
        programId: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
        keys: [
            { pubkey: vault, isSigner: false, isWritable: true },
            { pubkey: payer, isSigner: true, isWritable: true },
            { pubkey: mint, isSigner: false, isWritable: false },
            { pubkey: vaultAta, isSigner: false, isWritable: true },
            { pubkey: TOKEN_PROGRAM_ID, isSigner: false, isWritable: false },
            {
                pubkey: ASSOCIATED_TOKEN_PROGRAM_ID,
                isSigner: false,
                isWritable: false,
            },
            { pubkey: web3_js_1.SystemProgram.programId, isSigner: false, isWritable: false },
        ],
        data: Buffer.from([1, bump]),
    });
}
function transferToVaultIx(ephemeralAta, vault, mint, sourceAta, vaultAta, owner, amount) {
    return new web3_js_1.TransactionInstruction({
        programId: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
        keys: [
            { pubkey: ephemeralAta, isSigner: false, isWritable: true },
            { pubkey: vault, isSigner: false, isWritable: false },
            { pubkey: mint, isSigner: false, isWritable: false },
            { pubkey: sourceAta, isSigner: false, isWritable: true },
            { pubkey: vaultAta, isSigner: false, isWritable: true },
            { pubkey: owner, isSigner: true, isWritable: false },
            { pubkey: TOKEN_PROGRAM_ID, isSigner: false, isWritable: false },
        ],
        data: Buffer.from([2, ...u64le(amount)]),
    });
}
function depositSplTokensIx(ephemeralAta, vault, mint, sourceAta, vaultAta, owner, amount) {
    return transferToVaultIx(ephemeralAta, vault, mint, sourceAta, vaultAta, owner, amount);
}
function delegateEphemeralAtaIx(payer, ephemeralAta, bump, validator) {
    const data = validator
        ? Buffer.concat([Buffer.from([4, bump]), validator.toBuffer()])
        : Buffer.from([4, bump]);
    return new web3_js_1.TransactionInstruction({
        programId: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
        keys: [
            { pubkey: payer, isSigner: true, isWritable: true },
            { pubkey: ephemeralAta, isSigner: false, isWritable: true },
            {
                pubkey: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
                isSigner: false,
                isWritable: false,
            },
            {
                pubkey: (0, pda_js_1.delegateBufferPdaFromDelegatedAccountAndOwnerProgram)(ephemeralAta, constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID),
                isSigner: false,
                isWritable: true,
            },
            {
                pubkey: (0, pda_js_1.delegationRecordPdaFromDelegatedAccount)(ephemeralAta),
                isSigner: false,
                isWritable: true,
            },
            {
                pubkey: (0, pda_js_1.delegationMetadataPdaFromDelegatedAccount)(ephemeralAta),
                isSigner: false,
                isWritable: true,
            },
            { pubkey: constants_js_1.DELEGATION_PROGRAM_ID, isSigner: false, isWritable: false },
            { pubkey: web3_js_1.SystemProgram.programId, isSigner: false, isWritable: false },
        ],
        data,
    });
}
function initShuttleEphemeralAtaIx(payer, shuttleEphemeralAta, shuttleAta, shuttleWalletAta, owner, mint, shuttleId, bump) {
    if (!Number.isInteger(shuttleId) ||
        shuttleId < 0 ||
        shuttleId > 4294967295) {
        throw new Error("shuttleId must fit in u32");
    }
    const data = Buffer.alloc(6);
    data[0] = 11;
    data.writeUInt32LE(shuttleId, 1);
    data[5] = bump;
    return new web3_js_1.TransactionInstruction({
        programId: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
        keys: [
            { pubkey: payer, isSigner: true, isWritable: true },
            { pubkey: shuttleEphemeralAta, isSigner: false, isWritable: true },
            { pubkey: shuttleAta, isSigner: false, isWritable: true },
            { pubkey: shuttleWalletAta, isSigner: false, isWritable: true },
            { pubkey: owner, isSigner: false, isWritable: false },
            { pubkey: mint, isSigner: false, isWritable: false },
            { pubkey: TOKEN_PROGRAM_ID, isSigner: false, isWritable: false },
            {
                pubkey: ASSOCIATED_TOKEN_PROGRAM_ID,
                isSigner: false,
                isWritable: false,
            },
            { pubkey: web3_js_1.SystemProgram.programId, isSigner: false, isWritable: false },
        ],
        data,
    });
}
function delegateShuttleEphemeralAtaIx(payer, shuttleEphemeralAta, shuttleAta, bump, validator) {
    const data = validator
        ? Buffer.concat([Buffer.from([13, bump]), validator.toBuffer()])
        : Buffer.from([13, bump]);
    return new web3_js_1.TransactionInstruction({
        programId: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
        keys: [
            { pubkey: payer, isSigner: true, isWritable: true },
            { pubkey: shuttleEphemeralAta, isSigner: false, isWritable: false },
            { pubkey: shuttleAta, isSigner: false, isWritable: true },
            {
                pubkey: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
                isSigner: false,
                isWritable: false,
            },
            {
                pubkey: (0, pda_js_1.delegateBufferPdaFromDelegatedAccountAndOwnerProgram)(shuttleAta, constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID),
                isSigner: false,
                isWritable: true,
            },
            {
                pubkey: (0, pda_js_1.delegationRecordPdaFromDelegatedAccount)(shuttleAta),
                isSigner: false,
                isWritable: true,
            },
            {
                pubkey: (0, pda_js_1.delegationMetadataPdaFromDelegatedAccount)(shuttleAta),
                isSigner: false,
                isWritable: true,
            },
            { pubkey: constants_js_1.DELEGATION_PROGRAM_ID, isSigner: false, isWritable: false },
            { pubkey: web3_js_1.SystemProgram.programId, isSigner: false, isWritable: false },
        ],
        data,
    });
}
function mergeShuttleIntoAtaIx(owner, destinationAta, shuttleEphemeralAta, shuttleWalletAta, mint) {
    return new web3_js_1.TransactionInstruction({
        programId: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
        keys: [
            { pubkey: owner, isSigner: true, isWritable: false },
            { pubkey: destinationAta, isSigner: false, isWritable: true },
            { pubkey: shuttleEphemeralAta, isSigner: false, isWritable: false },
            { pubkey: shuttleWalletAta, isSigner: false, isWritable: true },
            { pubkey: mint, isSigner: false, isWritable: false },
            { pubkey: TOKEN_PROGRAM_ID, isSigner: false, isWritable: false },
        ],
        data: Buffer.from([15]),
    });
}
function undelegateAndCloseShuttleEphemeralAtaIx(payer, shuttleEphemeralAta, shuttleAta, shuttleWalletAta) {
    return new web3_js_1.TransactionInstruction({
        programId: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
        keys: [
            { pubkey: payer, isSigner: true, isWritable: true },
            { pubkey: shuttleEphemeralAta, isSigner: false, isWritable: false },
            { pubkey: shuttleAta, isSigner: false, isWritable: false },
            { pubkey: shuttleWalletAta, isSigner: false, isWritable: true },
            { pubkey: TOKEN_PROGRAM_ID, isSigner: false, isWritable: false },
            { pubkey: constants_js_1.MAGIC_CONTEXT_ID, isSigner: false, isWritable: true },
            { pubkey: constants_js_1.MAGIC_PROGRAM_ID, isSigner: false, isWritable: false },
        ],
        data: Buffer.from([14]),
    });
}
function withdrawSplIx(owner, mint, amount) {
    const [ephemeralAta] = deriveEphemeralAta(owner, mint);
    const [vault, vaultBump] = deriveVault(mint);
    const vaultAta = deriveVaultAta(mint, vault);
    const userDestAta = getAssociatedTokenAddressSync(mint, owner);
    return new web3_js_1.TransactionInstruction({
        programId: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
        keys: [
            { pubkey: owner, isSigner: true, isWritable: false },
            { pubkey: ephemeralAta, isSigner: false, isWritable: true },
            { pubkey: vault, isSigner: false, isWritable: false },
            { pubkey: mint, isSigner: false, isWritable: false },
            { pubkey: vaultAta, isSigner: false, isWritable: true },
            { pubkey: userDestAta, isSigner: false, isWritable: true },
            { pubkey: TOKEN_PROGRAM_ID, isSigner: false, isWritable: false },
        ],
        data: Buffer.from([3, ...u64le(amount), vaultBump]),
    });
}
function undelegateIx(owner, mint) {
    const userAta = getAssociatedTokenAddressSync(mint, owner);
    const [ephemeralAta] = deriveEphemeralAta(owner, mint);
    return new web3_js_1.TransactionInstruction({
        programId: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
        keys: [
            {
                pubkey: owner,
                isSigner: true,
                isWritable: false,
            },
            {
                pubkey: userAta,
                isSigner: false,
                isWritable: true,
            },
            {
                pubkey: ephemeralAta,
                isSigner: false,
                isWritable: false,
            },
            {
                pubkey: constants_js_1.MAGIC_CONTEXT_ID,
                isSigner: false,
                isWritable: true,
            },
            {
                pubkey: constants_js_1.MAGIC_PROGRAM_ID,
                isSigner: false,
                isWritable: false,
            },
        ],
        data: Buffer.from([5]),
    });
}
function createEataPermissionIx(ephemeralAta, payer, bump, flags = 0) {
    const permission = (0, pda_js_1.permissionPdaFromAccount)(ephemeralAta);
    return new web3_js_1.TransactionInstruction({
        programId: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
        keys: [
            { pubkey: ephemeralAta, isSigner: false, isWritable: true },
            { pubkey: permission, isSigner: false, isWritable: true },
            { pubkey: payer, isSigner: true, isWritable: true },
            { pubkey: web3_js_1.SystemProgram.programId, isSigner: false, isWritable: false },
            { pubkey: constants_js_1.PERMISSION_PROGRAM_ID, isSigner: false, isWritable: false },
        ],
        data: Buffer.from([6, bump, flags]),
    });
}
function resetEataPermissionIx(ephemeralAta, payer, bump, flags = 0) {
    const permission = (0, pda_js_1.permissionPdaFromAccount)(ephemeralAta);
    return new web3_js_1.TransactionInstruction({
        programId: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
        keys: [
            { pubkey: ephemeralAta, isSigner: false, isWritable: false },
            { pubkey: permission, isSigner: false, isWritable: true },
            { pubkey: payer, isSigner: true, isWritable: false },
            { pubkey: constants_js_1.PERMISSION_PROGRAM_ID, isSigner: false, isWritable: false },
        ],
        data: Buffer.from([9, bump, flags]),
    });
}
function delegateEataPermissionIx(payer, ephemeralAta, bump, validator) {
    const permission = (0, pda_js_1.permissionPdaFromAccount)(ephemeralAta);
    return new web3_js_1.TransactionInstruction({
        programId: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
        keys: [
            { pubkey: payer, isSigner: true, isWritable: true },
            { pubkey: ephemeralAta, isSigner: false, isWritable: true },
            { pubkey: constants_js_1.PERMISSION_PROGRAM_ID, isSigner: false, isWritable: false },
            { pubkey: permission, isSigner: false, isWritable: true },
            { pubkey: web3_js_1.SystemProgram.programId, isSigner: false, isWritable: false },
            {
                pubkey: (0, pda_js_1.delegateBufferPdaFromDelegatedAccountAndOwnerProgram)(permission, constants_js_1.PERMISSION_PROGRAM_ID),
                isSigner: false,
                isWritable: true,
            },
            {
                pubkey: (0, pda_js_1.delegationRecordPdaFromDelegatedAccount)(permission),
                isSigner: false,
                isWritable: true,
            },
            {
                pubkey: (0, pda_js_1.delegationMetadataPdaFromDelegatedAccount)(permission),
                isSigner: false,
                isWritable: true,
            },
            { pubkey: constants_js_1.DELEGATION_PROGRAM_ID, isSigner: false, isWritable: false },
            { pubkey: validator, isSigner: false, isWritable: false },
        ],
        data: Buffer.from([7, bump]),
    });
}
function undelegateEataPermissionIx(owner, ephemeralAta) {
    const permission = (0, pda_js_1.permissionPdaFromAccount)(ephemeralAta);
    return new web3_js_1.TransactionInstruction({
        programId: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
        keys: [
            { pubkey: owner, isSigner: true, isWritable: false },
            { pubkey: ephemeralAta, isSigner: false, isWritable: true },
            { pubkey: permission, isSigner: false, isWritable: true },
            { pubkey: constants_js_1.PERMISSION_PROGRAM_ID, isSigner: false, isWritable: false },
            { pubkey: constants_js_1.MAGIC_PROGRAM_ID, isSigner: false, isWritable: false },
            { pubkey: constants_js_1.MAGIC_CONTEXT_ID, isSigner: false, isWritable: true },
        ],
        data: Buffer.from([8]),
    });
}
async function delegateSpl(owner, mint, amount, opts) {
    const payer = opts?.payer ?? owner;
    const validator = opts?.validator;
    const initIfMissing = opts?.initIfMissing ?? true;
    const initVaultIfMissing = opts?.initVaultIfMissing ?? initIfMissing;
    const instructions = [];
    const [ephemeralAta, eataBump] = deriveEphemeralAta(owner, mint);
    const [vault, vaultBump] = deriveVault(mint);
    const vaultAta = deriveVaultAta(mint, vault);
    const ownerAta = getAssociatedTokenAddressSync(mint, owner);
    if (initIfMissing) {
        instructions.push(initEphemeralAtaIx(ephemeralAta, owner, mint, payer, eataBump));
    }
    if (initVaultIfMissing) {
        instructions.push(initVaultIx(vault, mint, payer, vaultBump), initVaultAtaIx(payer, vaultAta, vault, mint));
    }
    instructions.push(transferToVaultIx(ephemeralAta, vault, mint, ownerAta, vaultAta, owner, amount));
    instructions.push(delegateEphemeralAtaIx(payer, ephemeralAta, eataBump, validator));
    return instructions;
}
async function delegateSplIdempotent(owner, mint, amount, opts) {
    const payer = opts?.payer ?? owner;
    const validator = opts?.validator;
    const initVaultIfMissing = opts?.initVaultIfMissing ?? false;
    const initAtasIfMissing = opts?.initAtasIfMissing ?? false;
    const randomShuttleId = () => {
        const cryptoObj = globalThis?.crypto;
        if (cryptoObj?.getRandomValues !== undefined) {
            const buf = new Uint32Array(1);
            cryptoObj.getRandomValues(buf);
            return buf[0];
        }
        return Math.floor(Math.random() * 4294967296);
    };
    const shuttleId = opts?.shuttleId ?? randomShuttleId();
    const instructions = [];
    const [ephemeralAta, eataBump] = deriveEphemeralAta(owner, mint);
    const [vault, vaultBump] = deriveVault(mint);
    const vaultAta = deriveVaultAta(mint, vault);
    const ownerAta = getAssociatedTokenAddressSync(mint, owner);
    const [shuttleEphemeralAta, shuttleBump] = deriveShuttleEphemeralAta(owner, mint, shuttleId);
    const [shuttleAta, shuttleAtaBump] = deriveShuttleAta(shuttleEphemeralAta, mint);
    const shuttleWalletAta = deriveShuttleWalletAta(mint, shuttleEphemeralAta);
    if (initVaultIfMissing) {
        instructions.push(initVaultIx(vault, mint, payer, vaultBump), initVaultAtaIx(payer, vaultAta, vault, mint));
    }
    if (initAtasIfMissing) {
        instructions.push(createAssociatedTokenAccountIdempotentInstruction(payer, ownerAta, owner, mint), createAssociatedTokenAccountIdempotentInstruction(payer, shuttleWalletAta, shuttleEphemeralAta, mint));
    }
    instructions.push(initEphemeralAtaIx(ephemeralAta, owner, mint, payer, eataBump), createEataPermissionIx(ephemeralAta, payer, eataBump), delegateEphemeralAtaIx(payer, ephemeralAta, eataBump, validator), initShuttleEphemeralAtaIx(payer, shuttleEphemeralAta, shuttleAta, shuttleWalletAta, owner, mint, shuttleId, shuttleBump));
    if (amount > 0n) {
        instructions.push(depositSplTokensIx(shuttleAta, vault, mint, ownerAta, vaultAta, owner, amount));
    }
    instructions.push(delegateShuttleEphemeralAtaIx(payer, shuttleEphemeralAta, shuttleAta, shuttleAtaBump, validator));
    return instructions;
}
async function delegatePrivateSpl(owner, mint, amount, opts) {
    const payer = opts?.payer ?? owner;
    const validator = opts?.validator ?? DEFAULT_PRIVATE_VALIDATOR;
    const initIfMissing = opts?.initIfMissing ?? true;
    const initVaultIfMissing = opts?.initVaultIfMissing ?? false;
    const permissionFlags = opts?.permissionFlags ?? 0;
    const delegatePermission = opts?.delegatePermission ?? false;
    const instructions = [];
    const [ephemeralAta, eataBump] = deriveEphemeralAta(owner, mint);
    const [vault, vaultBump] = deriveVault(mint);
    const vaultAta = deriveVaultAta(mint, vault);
    const ownerAta = getAssociatedTokenAddressSync(mint, owner);
    if (initIfMissing) {
        instructions.push(initEphemeralAtaIx(ephemeralAta, owner, mint, payer, eataBump));
    }
    if (initVaultIfMissing) {
        instructions.push(initVaultIx(vault, mint, payer, vaultBump));
        instructions.push(initVaultAtaIx(payer, vaultAta, vault, mint));
    }
    instructions.push(transferToVaultIx(ephemeralAta, vault, mint, ownerAta, vaultAta, owner, amount));
    instructions.push(delegateEphemeralAtaIx(payer, ephemeralAta, eataBump, validator));
    instructions.push(createEataPermissionIx(ephemeralAta, payer, eataBump, permissionFlags));
    if (delegatePermission) {
        instructions.push(delegateEataPermissionIx(payer, ephemeralAta, eataBump, validator));
    }
    return instructions;
}
function u64le(n) {
    if (n < 0n || n > 0xffffffffffffffffn) {
        throw new Error("amount out of range for u64");
    }
    const bytes = new Array(8).fill(0);
    let x = n;
    for (let i = 0; i < 8; i++) {
        bytes[i] = Number(x & 0xffn);
        x >>= 8n;
    }
    return bytes;
}
